### ISSUE TITLE
No Internet Connection When Using `exit-node` on Headscale

### ISSUE THREAD

**Issue by berrabe** [tailscale/tailscale#9199]
### What is the issue?

# TLDR
I used `headscale` with `docker` instead of `tailscale`.  Everything work flawlessly except i can't use `exit-node` using this [`tailscale CLI`](https://github.com/tailscale/tailscale) client on my Mac and Linux (Arch Linux), when the `exit-node` is active, I can't access the internet at all

But when I use the tailscale mobile application on iOS, the `exit-node` can run as expected to route all traffic. There is no difference in settings, including the ACL between the IOS application and the tailscale CLI users / namespace

### Steps to reproduce

- advertise `exit-node` on my public cloud VM (Linux)
```bash
tailscale set --advertise-exit-node
```

- accept routes on `headsacale` server (`0.0.0.0/0`) exposed by advertised node
```bash
docker exec -it headscale headscale routes list
docker exec -it headscale headscale routes enable -r <id>
```

- apply on my local client (MacOS)
```bash
tailscale exit-node list
tailscale set --exit-node=<my machine hostname or ipv4>
```

- with this [`tailscale CLI`](https://github.com/tailscale/tailscale) the route table not applied automatically, so i need to do it manual with restarting the tailscale
```bash
tailscale down
tailscale up
```

- but after the command above, my Laptop won't connect to internet at all ... I need to disable `exit-node` and manually add def route on my Mac
```
sudo route add default <gateway>
```

### OS

Linux, macOS, iOS

### OS version

MacOS Monterey 12.6.6

### Tailscale version

tailscale 1.48.1 on all devices (MacOS, Linux Cloud VM)

**Comment by juanfont** [comment]
@berrabe Headscale dev here. This is not a Tailscale bug.

Please check our Discord community. Also, make sure you have IP forwarding enabled in your exit node.

**Comment by berrabe** [comment]
@juanfont I don't know what causes this issue exactly, but as I explained above, when using the IOS app `exit-node` works, but when using the `tailscale CLI` `exit-node` doesn't work. There is no difference in settings between the two systems, using the same ACL and user / namespace

And I had asked to the `headscsale` discord community, and tried several methods from the community but still `exit-node` won't work

**Comment by juanfont** [comment]
In that case that'd probably be an issue somewhere else - perhaps in Headscale itself.

**Comment by berrabe** [comment]
## UPDATE

- I've try account from `tailscale.com` not `headscale`. And the result is same, `exit-node` still won't work on my Mac

![image](https://github.com/tailscale/tailscale/assets/34432326/561624d6-06f2-43c9-9be1-c6a0e11b65ac)

- need to restart `tailscale` manually after set `exit-node`. And my Mac become disconnected from the internet

![image](https://github.com/tailscale/tailscale/assets/34432326/4fa00eaa-f9b7-43e7-87a8-010906a5c15d)

- This is my ACL (default) on https://login.tailscale.com/admin/acls/file
```json
"acls": [
  {"action": "accept", "src": ["*"], "dst": ["*:*"]},
]
```

- And this is my `sysctl.conf`
```bash
net.ipv4.conf.all.forwarding = 1
net.ipv4.ip_forward = 1
net.ipv6.conf.all.forwarding = 1
```

**Comment by berrabe** [comment]
> While the Mac is trying to use the exit-node, please hold Option while clicking on the Tailscale icon in the menubar and select Debug -> Bug report. Please paste the bugreport here, so we can look into the issue.

@DentonGentry  i'm not using tailscale GUI (brew casks or app store app in this case) but this tailscale CLI which doesn't have the GUI ... and when running `tailscale bugreport`, the program return `BUG-NO-LOGS-NO-SUPPORT-this-node-has-had-its-logging-disabled`, i'm not passing `--no-logs-no-support` to the `tailscaled` daemon

**Comment by wazhanudin** [comment]
I have the same issue.

Exit node = VPS Debian 12 - 1.52.1
Use exit node from Local Windows 10 - 1.52.1 (No problem)
Use exit node from Local Android 13 - 1.52.0 (No problem)
Use exit node from Local Debian 12 - 1.52.1 (no internet) [cli]

`tailscale set --exit-node=1xx.1xx.1xx.1xx --exit-node-allow-lan-access=true`

I can reach all available lan from Local Debian 12 - 1.52.1
I can't reach my VPS Debian 12 - 1.52.1 from Local Debian 12 - 1.52.1

**Comment by j0nathontayl0r** [comment]
Seeing a similar issue as @wazhanudin and @berrabe

Edit node works on all other clients with Tailscale, tested on Linux and Windows, but not on MacOS. 

Here's a bug report from me @DentonGentry

BUG-6bf898e594a8f7857783644f5291888f04e807d1c67647fb488c2096ca9407e9-20231230110347Z-f399ef02d2c74c35

Disabling the MacOS firewall hasn't changed anything, and disabling iptables on the Linux Exit Node hasn't helped. Happy to help troubleshoot further.

**Comment by j0nathontayl0r** [comment]
Searched through some of the existing and closed issues to help troubleshoot. 

1. DNS is working ok, tested with `dig google.com +short` amongst others
2. Internet isn't, confirmed with `curl https://1.1.1.1` and `curl https://google.com` amongst others, both fail
3. Looked into iptables settings on the Debian Exit Node, allowing all ingress and egress didn't seem to have any material impact
4. I can hit the Debian 11 server just fine on SSH, `ping $TailscaleExitNodeName` and `tailscale ping`$TailscaleExitNodeName` without issues
5. Changed my DNS setup to not be using NextDNS and exclusively use CloudFlare DNS, no changes
6. tcpdump shows my MacOS laptop is hitting the exit node, but connections aren't getting back out (log attached, [tcpdump.log](https://github.com/tailscale/tailscale/files/13797896/tcpdump.log))

**Comment by thetechyghost** [comment]
Had the exact same issue until I updated my server to the last version (1.56.x - I think). Working fine now.

**Comment by mrafayaleem** [comment]
Is there any chance that your exit node is using a local dns server on your LAN such as AdGuard? If that's the case, try assigning static DNS IP such as `8.8.8.8` or `1.1.1.1` to your exit node and then try and see if that works.

That is what made it work for me for my exit node runnning on Unraid.

**Comment by thetechyghost** [comment]
> Is there any chance that your exit node is using a local dns server on your LAN such as AdGuard? If that's the case, try assigning static DNS IP such as `8.8.8.8` or `1.1.1.1` to your exit node and then try and see if that works.
> 
> 
> 
> That is what made it work for me for my exit node runnning on Unraid.

I do use a local DNS server, but this is a requirement in my configuration.

**Comment by immanwell** [comment]
@thetechyghost  Did you find a solution to this because im having the same problem and my setup requires to use local dns for all tailscale connected devices fro use adguard home.

**Comment by thetechyghost** [comment]
> [@thetechyghost](https://github.com/thetechyghost) Did you find a solution to this because im having the same problem and my setup requires to use local dns for all tailscale connected devices fro use adguard home.

Updating my software to the latest version fixed it and has been smooth sailing ever since.

**Comment by whrami** [comment]
Not for me. I have linux boxes, Ubuntu, debian, busybox, openbsd, and others. All of them work with my exit node, but my Macbooks do not.

**Comment by leetdavid** [comment]
~In case anyone has the same issue as me, this is what worked for me:~ (update: it doesn't work consistently)

1. Go to the DNS tab in your tailscale admin page (https://login.tailscale.com/admin/dns).
2. Add Cloudflare Public DNS (1.1.1.1).
3. Check "Override DNS Servers"
4. Click on the 3 dots and click "Edit...", and turn on "Use with exit node". It should look like this: 

<img width="736" height="328" alt="Image" src="https://github.com/user-attachments/assets/877d76b8-8e88-4c5d-9b4a-e05c75221886" />

5. In your mac's Tailscale settings, toggle "Use Tailscale DNS settings" to update the settings on your client machine.

UPDATE: it only worked for my macbook, but not my mac mini which I use as a dev server. guess i'll go back to the drawing board...
